### ISSUE TITLE
Error Reporting is not Ignoring Classes

### ISSUE THREAD

**Issue by dshimy** [googleapis/google-cloud-ruby#2605]
#### Environment details

  - OS: GKE/Debian (Stretch)
  - Rails: 5.2.1
  - Ruby version: 2.4.5
  - Gem name and version: stackdriver (0.15.3) 

#### Steps to reproduce

We want to ignore the following errors: ActiveRecord::RecordNotFound, ActionController::RoutingError.  This is not something that broke recently, we never got it to work.

1. Edit the production.rb (see below) file to ignore the classes.
2. Visit URL that does not exist.  Error is still triggered.

Here is the log entry (note the exception in the textPayload that doesn't wrap):

```json
[
 {
   "textPayload": "actionpack (5.2.1) lib/action_dispatch/middleware/debug_exceptions.rb:65:in `call'\nactionpack (5.2.1) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'\nrailties (5.2.1) lib/rails/rack/logger.rb:38:in `call_app'\nrailties (5.2.1) lib/rails/rack/logger.rb:28:in `call'\ngoogle-cloud-logging (1.5.6) lib/google/cloud/logging/middleware.rb:91:in `call'\nactionpack (5.2.1) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'\nrequest_store (1.4.1) lib/request_store/middleware.rb:19:in `call'\nactionpack (5.2.1) lib/action_dispatch/middleware/request_id.rb:27:in `call'\nrack (2.0.6) lib/rack/method_override.rb:22:in `call'\nrack (2.0.6) lib/rack/runtime.rb:22:in `call'\ngoogle-cloud-trace (0.33.5) lib/google/cloud/trace/middleware.rb:157:in `block in call'\ngoogle-cloud-trace (0.33.5) lib/google/cloud/trace.rb:258:in `block in in_span'\ngoogle-cloud-trace (0.33.5) lib/google/cloud/trace/trace_record.rb:232:in `in_span'\ngoogle-cloud-trace (0.33.5) lib/google/cloud/trace.rb:255:in `in_span'\ngoogle-cloud-trace (0.33.5) lib/google/cloud/trace/middleware.rb:155:in `call'\nactivesupport (5.2.1) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'\nactionpack (5.2.1) lib/action_dispatch/middleware/executor.rb:14:in `call'\nactionpack (5.2.1) lib/action_dispatch/middleware/static.rb:127:in `call'\nrack (2.0.6) lib/rack/sendfile.rb:111:in `call'\nrack-cors (1.0.2) lib/rack/cors.rb:97:in `call'\nrailties (5.2.1) lib/rails/engine.rb:524:in `call'\nprometheus-client (0.8.0) lib/prometheus/middleware/exporter.rb:32:in `call'\nprometheus-client (0.8.0) lib/prometheus/middleware/collector.rb:38:in `block in call'\nprometheus-client (0.8.0) lib/prometheus/middleware/collector.rb:84:in `block in trace'\n/usr/local/lib/ruby/2.4.0/benchmark.rb:308:in `realtime'\nprometheus-client (0.8.0) lib/prometheus/middleware/collector.rb:84:in `trace'\nprometheus-client (0.8.0) lib/prometheus/middleware/collector.rb:38:in `call'\npuma (3.12.0) lib/puma/configuration.rb:225:in `call'\npuma (3.12.0) lib/puma/server.rb:658:in `handle_request'\npuma (3.12.0) lib/puma/server.rb:472:in `process_client'\npuma (3.12.0) lib/puma/server.rb:332:in `block in run'\npuma (3.12.0) lib/puma/thread_pool.rb:133:in `block in spawn_thread'",
   "insertId": "h7ky2of3mb1vd",
   "resource": {
     "type": "container",
     "labels": {
       "cluster_name": "prod-cluster",
       "container_name": "",
       "namespace_id": "default",
       "instance_id": "",
       "zone": "",
       "pod_id": "",
       "project_id": "PROJECT"
     }
   },
   "timestamp": "2018-11-15T15:03:40.752366668Z",
   "severity": "CRITICAL",
   "labels": {
     "traceId": "a3446c718cd7463682110b28642736e2"
   },
   "logName": "projects/PROJECT/logs/app_log",
   "trace": "projects/PROJECT/traces/a3446c718cd7463682110b28642736e2",
   "receiveTimestamp": "2018-11-15T15:03:40.932012815Z"
 }
]
```

#### Code example

```ruby
# config/environments/production.rb
Rails.application.configure do
  config.google_cloud.error_reporting.ignore_classes = [ActiveRecord::RecordNotFound, ActionController::RoutingError]
end
```

**Comment by blowmage** [comment]
Hi, I'm looking into this and I'm a little confused. You are showing a Logging log entry, but you are asking about Error Reporting. The `ignore_classes` configuration is for the ErrorReporting service, not the Logging service. Are you asking that these requests not be logged?

**Comment by blowmage** [comment]
@dshimy Can you clarify the behavior you are expecting to see here?

**Comment by blowmage** [comment]
@dshimy Ping.

**Comment by dshimy** [comment]
@blowmage Sorry for the delay and confusion.  

What we want to do is prevent ActiveRecord::RecordNotFound and ActionController::RoutingError errors from showing up in the GCP Error Reporting interface.

From the documentation, there is an `error_reporting.ignore_classes` option on this page: https://github.com/googleapis/google-cloud-ruby/blob/master/stackdriver/INSTRUMENTATION_CONFIGURATION.md#error-reporting

Fully appreciation I might be missing something here.

**Comment by blowmage** [comment]
What version of google-cloud-error_reporting are you using?

**Comment by blowmage** [comment]
Are you also setting `ignore_classes` on `Google::Cloud::ErrorReporting.configure` anywhere else in your app? The runtime checks for values at `Google::Cloud::ErrorReporting.configure.ignore_classes` before checking `Rails.application.config.google_cloud.error_reporting.ignore_classes`.

**Comment by blowmage** [comment]
Are you doing anything extra, like calling `Google::Cloud::ErrorReporting.report` somewhere? What I'm seeing is that by default Rails already handles `ActiveRecord::RecordNotFound` and `ActionController::RoutingError` and those errors are not reported to ER. I have not been able to get either of those errors to show up on the Error Reporting dashboard.

I should note that `ignore_classes` is only for errors that are reported by the Rack Middleware. Any error you pass to `Google::Cloud::ErrorReporting.report` will be reported, even if they match a class listed in `ignore_classes`.

**Comment by dshimy** [comment]
Here are the answers to your questions:

1. The version is `google-cloud-error_reporting (0.30.5)`
2. We are not setting it again.  Here is the config, just two lines:

        config.google_cloud.error_reporting.service_name = "MyApp"
        config.google_cloud.error_reporting.ignore_classes = [ActiveRecord::RecordNotFound, ActionController::RoutingError]

3. I searched the code for `RecordNotFound` to see if we had any special handling and we don't.

Ironically, I checked to see if the problem is still happening and it's not.  The only change we made was to disable Stackdriver logging.  We also added a Prometheus exporter to Rack, but I don't think that would change what I was experiencing.  

So for my needs, I'm good now, but there may still be an issue between logging and error reporting.
